loopy/__init__.py

```diff
@@ -487,6 +487,48 @@ def make_copy_kernel(new_dim_tags, old_dim_tags=None):
 # }}}
 
 
+# {{{ einsum
+
+def make_einsum(spec, arg_names, knl_creation_kwargs=None):
+    arg_spec, out_spec = spec.split("->")
+    arg_specs = arg_spec.split(",")
+
+    out_indices = set(out_spec)
+
+    all_indices = set(
+        idx
+        for argsp in arg_specs
+        for idx in argsp) | out_indices
+
+    sum_indices = all_indices - out_indices
+
+    from pymbolic import var
+    lhs = var("out")[tuple(var(i) for i in out_spec)]
+
+    rhs = 1
+    for arg_name, argsp in zip(arg_names, arg_specs):
+        rhs = rhs * var(arg_name)[tuple(var(i) for i in argsp)]
+
+    if sum_indices:
+        rhs = Reduction("sum", tuple(var(idx) for idx in sum_indices), rhs)
+
+    constraints = " and ".join(
+        "0 <= %s < N%s" % (idx, idx)
+        for idx in all_indices
+        )
+
+    if "name" not in knl_creation_kwargs:
```

Comment:
This will break if `knl_creation_kwargs is None`. Should we add something like
```python
knl_creation_kwargs = knl_creation_kwargs or {}
```
above? Or should we just pass `**kwargs` to `make_einsum`?

Reply:
I prefer dictionary unpacking :)

Reply:
I feel that if we pass on `**kwargs`, it may be difficult to extend this interface later. (because `make_kernel` kwargs and kwargs for this share a namespace)

Reply:
OTOH, maybe passing on `**kwargs` is not so bad.

Reply:
How's this https://github.com/inducer/loopy/pull/328/commits/c44d8fed5e5a5115ea8e61f92b5cc60f22b9fed0?

Reply:
:+1: